gem/optimise.py

```diff
@@ -413,3 +474,403 @@ def aggressive_unroll(expression):
     expression, = unroll_indexsum((expression,), max_extent=numpy.inf)
     expression, = remove_componenttensors((expression,))
     return expression
+
+
+@singledispatch
+def count_flop_node(node):
+    """Count number of flops at a particular gem node, without recursing
+    into childrens"""
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+@count_flop_node.register(Constant)
+@count_flop_node.register(Terminal)
+@count_flop_node.register(Indexed)
+@count_flop_node.register(Variable)
+@count_flop_node.register(ListTensor)
+@count_flop_node.register(FlexiblyIndexed)
+@count_flop_node.register(IndexSum)
+@count_flop_node.register(LogicalNot)
+@count_flop_node.register(LogicalAnd)
+@count_flop_node.register(LogicalOr)
+@count_flop_node.register(Conditional)
+def count_flop_node_zero(node):
+    return 0
+
+
+@count_flop_node.register(Power)
+@count_flop_node.register(Comparison)
+@count_flop_node.register(Sum)
+@count_flop_node.register(Product)
+@count_flop_node.register(Division)
+@count_flop_node.register(MathFunction)
+def count_flop_node_single(node):
+    return numpy.prod([idx.extent for idx in node.free_indices])
+
+
+def count_flop(node):
+    """Count total number of flops of a gem expression, assuming hoisting and
+    reuse"""
+    flops = sum(map(count_flop_node, traversal([node])))
+
+    return flops
```

Comment:
Amended.
Some simple tests together with optimisation. Will think of a few more.